massa-execution-worker/src/speculative_roll_state.rs

```diff
@@ -298,51 +298,96 @@ impl SpeculativeRollState {
+                        .pos_changes
+                        .deferred_credits
+                        .credits
+                        .range(min_slot..max_slot)
+                    {
+                        if let Some(amount) = addr_amount.get(address) {
+                            res.entry(*slot).or_insert(*amount);
+                        };
+                    }
+                }
+            }
+
+            // get values from final state
+            {
+                let final_state = self.final_state.read();
+                let deferred_credits = final_state
+                    .get_pos_state()
+                    .get_deferred_credits_range(min_slot..max_slot, Some(*address));
+                for (slot, addr_amount) in deferred_credits.credits {
                     if let Some(amount) = addr_amount.get(address) {
-                        res.entry(*slot).or_insert(*amount);
+                        res.entry(slot).or_insert(*amount);
                     };
                 }
             }
-        }
-
-        // get values from final state
-        {
-            let final_state = self.final_state.read();
-            for (slot, addr_amount) in final_state
-                .get_pos_state()
-                .get_deferred_credits_range(min_slot..)
+        } else {
+            // get added values
+            for (slot, addr_amount) in self
+                .added_changes
+                .deferred_credits
                 .credits
+                .range(min_slot..)
             {
                 if let Some(amount) = addr_amount.get(address) {
-                    res.entry(slot).or_insert(*amount);
+                    res.entry(*slot).or_insert(*amount);
                 };
             }
+
+            // get values from active history, backwards
+            {
+                let hist = self.active_history.read();
+                for hist_item in hist.0.iter().rev() {
+                    for (slot, addr_amount) in hist_item
+                        .state_changes
+                        .pos_changes
+                        .deferred_credits
+                        .credits
+                        .range(min_slot..)
+                    {
+                        if let Some(amount) = addr_amount.get(address) {
+                            res.entry(*slot).or_insert(*amount);
+                        };
+                    }
+                }
+            }
+
+            // get values from final state
+            {
+                let final_state = self.final_state.read();
+                let deferred_credits = final_state
+                    .get_pos_state()
+                    .get_deferred_credits_range(min_slot.., Some(*address));
+                for (slot, addr_amount) in deferred_credits.credits {
+                    if let Some(amount) = addr_amount.get(address) {
+                        res.entry(slot).or_insert(*amount);
+                    };
+                }
+            }
```

Comment:
Maybe I'm missing something, but why can't you factor this code with a:
```rust
let range = if let Some(max_slot) = max_slot_ {
  RangeBound(Included(min_slot), Excluded(max_slot))
} else {
  RangeBound(Included(min_slot), Unbounded))
}
```

Reply:
RangeBounds is a trait and not a type (https://doc.rust-lang.org/std/ops/trait.RangeBounds.html) no?

Reply:
Haven't tested it myself so I trust you :)

Reply:
Seems to be just Bound to have an object apparently :) https://doc.rust-lang.org/std/ops/enum.Bound.html